### ISSUE TITLE
Code base Rewamp

### ISSUE THREAD

**Issue by Sauilitired** [IntellectualSites/PlotSquared#1805]
The current state of the code base is quite poor. It is very fragmented, and there is the whole issue with split packaging. There is also a lack of consistency in the code style. This is very hard to solve without breaking compatibility with plugins relying on the PlotSquared API. But, as this has been discussed for years now, I feel as if fixing these issues would justify breaking compatibility. 

The following issues needs to be addressed:
- Code is split up into multiple packages. I suggest we use "com.github.intellectualsites.plotsquared", which follows the Maven Central guidelines for namespaces. 
- A code style should be adopted and enforced. I suggest using the code style from [Kvantum](https://github.com/IntellectualSites/Kvantum), however I am open to suggestions.
- As of the recent news that material IDs will be dropped completely in Minecraft 1.13, the project should move to material enums. (See #1795). A custom solution should be implemented, that will keep support for both vanilla materials and modded items. 

I would also like to hear opinions on using [lombok](http://projectlombok.org) throughout PlotSquared, which would allow for much cleaner code, with less boilerplate. This is a well established library and would only need to be included as a compile time dependency. 

These issues would be addressed in a separate branch as of now (remake) until the project is functioning properly. Important fixes would still be pushed onto the master branch during the "remake". When the process is finished, the master branch would move into a legacy branch and remake would overwrite master. Thus server owners would still be able to use the plugin until their dependencies are updated. 

I would like to hear your opinions on this.

**Comment by Opachl** [comment]
Hi can i join this Project. To help you out.?

**Comment by Sauilitired** [comment]
@boy0001 my English is getting worse by the day, thanks for helping me :p

@Opachl yes! All contributions are appreciated

**Comment by Opachl** [comment]
How did we organize us? Did i need to join somewhere?

**Comment by RedstoneFuture** [comment]
I can help with testing. I work with @Opachl together.

**Comment by Sauilitired** [comment]
@Opachl and @RedstoneFuture I'm going on my Christmas break next week, so I'll get started then. I'll figure out an efficient way of communicating. Although, I guess Discord is probably the best option?

**Comment by Opachl** [comment]
Sounds good. :) and something to organize our substeps i would suggest something like trello.

**Comment by lurr** [comment]
If we're going to restructure PlotSquared; we should also try to update and fix some issues with the Sponge builds. I'd be willing to do a bit of work there.

**Comment by Sauilitired** [comment]
@lurr Definitely. I have very little experience with sponge myself, so I've been scared to work with the sponge implementation.

**Comment by RedstoneFuture** [comment]
@Sauilitired HI: would you like add me on discord: RedstoneFuture#3356

**Comment by Opachl** [comment]
@Sauilitired  My discord is Alina Ealasaid#4687

**Comment by Opachl** [comment]
Hi i just wanted to ask the state of this ;) for personal planing

**Comment by Sauilitired** [comment]
I have done quite a lot of refactoring locally; but I don't want to push until it is at least somewhat working.

I'll be quite honest, I have both school and jobs to take care of, so this is not a big priority of mine right now.
________________________________

**Comment by RedstoneFuture** [comment]
OK. I unterstand and I have a idea: if you work to the point that we can help, i think the big rest is only communication (fast and effective). (In our project planing we have set the priority to higt.) :)

**Comment by lurr** [comment]
Discord : Lur#2961
A SpongeAPI 7.0.0/7.1.0 update is in order

**Comment by Narimm** [comment]
Does the plan here involve actually sorting the mess that gradle is creating ... I am not buying into the whole maven v gradle thing - but the reality is even gradle uses maven to pull deps...and currently there is a massive mismatch between maven versioning the plugin version and even build version.  I mean the current gradle build outputs a version based around the the git log and commit date  as best I can work out.  I use plotsquared as a project dep with a deeper integration than the api allows... but the fact it doesnt get pushed to a public repo so it can be pulled as dependency is annoying.

**Comment by boy0001** [comment]
@Narimm
No. Gradle allows for more control, and is required to build the sponge module.

The PlotSquared-API jar has a maven repository:
https://github.com/IntellectualSites/PlotSquared/wiki/Developer-documentation-(WIP)#maven

**Comment by Opachl** [comment]
Any updates here ? :)

**Comment by boy0001** [comment]
#### Bundle a blocks.json which maps the old ids to the block names:
Similar to what WorldEdit does currently see [here](https://raw.githubusercontent.com/sk89q/WorldEdit/master/worldedit-core/src/main/resources/com/sk89q/worldedit/world/registry/blocks.json)

#### IPlotMain
 - getBlockRegistry()

#### BlockRegistry (interface)
 - register(string id, int legacyId)
 - getId(int name)
 - getName(char id)
    
#### PlotBlock
 > Should become an abstract class (or interface?), which is implemented by the platform

 - static get(int id)
 - static get(string variant)
 
#### BukkitPlotBlock
 - Material variant;
    
#### During startup:
 - Register every variant in the blocks.json with the registry (which should have all legacy blocks)
 - Register every other variant present on the server (for bukkit this would be from the Material class)
